openedx/core/lib/gating/api.py

```diff
@@ -299,3 +304,112 @@ def get_gated_content(course, user):
                 {'id': user.id}
             )
         ]
+
+
+def is_prereq_met(content_id, user_id, recalc_on_unmet=False):
+    """
+    Returns true if the prequiste has been met for a given milestone
+
+    Arguments:
+        content_id (BlockUsageLocator): BlockUsageLocator for the content
+        user_id: The id of the user
+        recalc_on_unmet: Recalculate the grade if prereq has not yet been met
+
+    Returns:
+        tuple: True|False,
+        prereq_meta_info = { 'url': prereq_url|None, 'display_name': prereq_name|None}
+    """
+    course_id = content_id.course_key
+
+    # if unfullfilled milestones exist it means prereq has not been met
+    unfulfilled_milestones = milestones_helpers.get_course_content_milestones(
+        course_id,
+        content_id,
+        'requires',
+        user_id
+    )
+    prereq_met = not unfulfilled_milestones
+    prereq_meta_info = {'url': None, 'display_name': None}
+
+    if prereq_met or not recalc_on_unmet:
+        return prereq_met, prereq_meta_info
+
+    milestone = unfulfilled_milestones[0]
+    student = User.objects.get(id=user_id)
+    store = modulestore()
+
+    with store.bulk_operations(course_id):
+        course_structure = get_course_blocks(student, store.make_course_usage_key(course_id))
+        course = store.get_course(course_id, depth=0)
+        subsection_grade_factory = SubsectionGradeFactory(student, course, course_structure)
+        subsection_usage_key = UsageKey.from_string(_get_gating_block_id(milestone))
+
+        if subsection_usage_key in course_structure:
+            # this will force a recalcuation of the subsection grade
+            subsection_grade = subsection_grade_factory.update(course_structure[subsection_usage_key])
```

Comment:
I haven't thought about this - but do we want this call to have the side-effect of also persisting the newly computed grade?  By default, this call will persist the new value.  There is logic for not persisting, which we could activate if we choose not to persist.

Reply:
@nasthagiri 
If we didn't persist it, would the Progress page not reflect the newly calculated grade? I think we'd want the Progress page to show correct grade, as that has been one complaint from learners that the Progress page doesn't update in a timely manner. This would at least help that if you visit the gated page. Are there other side-effects/race conditions of potentially updating twice?

I also had the same question about persisting the milestone record to the database in the next line call to update_milestone. @ormsbee discussed a bit and thought it might be safest to update the db to avoid any inconsistencies that might arise from in-memory and db being out of sync.

Reply:
This would be the first instance where we update the subsection grade - but not update the course grade.  This may cause confusion, unless:

**Option 1:** Update the entire course grade instead - calling CourseGradeFactory.update - taking the performance hit - having the learner wait - the request may also timeout since this is in a web worker

**Option 2:** Calculate, but don't persist, the new Subsection grade value - should not timeout. Since we don't persist, there should be no inconsistency on the progress page (i.e., the subsection grades will add up to the course grade)

**Option 3:** Kickoff the regrade and have a web-socket channel update the UI

other options?

Reply:
The issue with the progress page not updating in a timely manner is being tackled by the Educator team - at least initially via UI changes so the learner sees the last-modified timestamp.  FYI @sstack22 - to confirm that's happening.

Reply:
@nasthagiri @ormsbee 
I would vote for Option 2 for the short term, with eyes on Option 3 for future. The Pros being it enables us to displayed gated sections immediately to user if they meet the prereq, which is a big win and the main problem we set out to solve here. Con (if I understand correctly) is the Progress page wouldn't necessarily reflect the forced Subsection grade right away, but that seems like an acceptable tradeoff.

If we go with Option 2, should I add another argument to SubsectionGradeFactory.update() to prevent persisting? Looks like the current mechanism is checking a feature flag on the course  to determine whether to persist the grade or not.

Reply:
Yes.  Adding an argument for persisting, defaulting to `True` to `SubsectionGradeFactory.update()` would be needed for Option 2.

Reply:
@nasthagiri 
Done. Was a bit confused in my testing though. Couple of follow on questions:
1) Looks like PersistentGradesEnabledFlag.feature_enabled(course_key) is false in my devstack so never trying to persist for me even without the flag. How do I enable this?
2) Even without persisting the grade, the Progress page seemed to show the correct scores for the problems. Was curious where it was reading the grades from, guessing in memory somewhere.
3) What's the best way to manually test a blocked/slow grading queue? I no-op'd _update_subsection_grades() in lms/djangoapps/grades/tasks.py but was surprised to see Progress page still updating after changing answers to some problems.

Reply:
Sorry for the confusion.
1. You can enable this via Django admin by going to localhost:18000/admin/grades/persistentgradesenabledflag/
2. When persistent grades is not enabled, we always recompute the grade whenever you visit the Progress page.  This was the default behavior before we implemented Persistent grades.
3. Yeah.  It's because visiting the Progress page always recomputes the grade if Persistent grades isn't enabled.  Once you enable it, you should see the expected behavior.